# Exploring Ebay Car Sales Data

For this project we will analyze a dataset of used cars from *Ebay Kleinanzeigen*, a classifieds sections of the German eBay website.

The aim of this project is to clean and analyze the data of used car listing using Pandas.

In [263]:
import numpy as np
import pandas as pd
from csv import reader

In [264]:
autos=pd.read_csv('autos.csv',encoding='Latin-1')

In [265]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [266]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [267]:
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

From the information extracted we can notice that:
* There are 50000 entries in each of the 20 columns, 15 of them are strings and the rest are integers.
* There are null values in the `vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage` columns.
* The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

### Cleaning Column Names

We will convert the column names from camelcase to snakecase and reword some of the columns names based on the data dictionary to be more decriptive

In [268]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [269]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Data Exploration

In [270]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

In [271]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                 object
ab_test               object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer              object
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
num_photos             int64
postal_code            int64
last_seen             object
dtype: object

In [272]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [273]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [274]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

In [275]:
for c in autos:
    value=autos[c].value_counts()
    print(value)

2016-03-29 23:42:13    3
2016-03-08 10:40:35    3
2016-03-23 18:39:34    3
2016-03-23 19:38:20    3
2016-03-27 22:55:05    3
2016-03-21 16:37:21    3
2016-03-12 16:06:22    3
2016-03-21 20:37:19    3
2016-04-02 11:37:04    3
2016-03-16 21:50:53    3
2016-03-22 09:51:06    3
2016-03-25 19:57:10    3
2016-03-11 22:38:16    3
2016-03-10 15:36:24    3
2016-03-14 20:50:02    3
2016-04-04 16:40:33    3
2016-04-02 15:49:30    3
2016-03-05 16:57:05    3
2016-03-30 17:37:35    3
2016-03-30 19:48:02    3
2016-03-09 11:54:38    3
2016-03-19 17:36:18    3
2016-03-17 20:40:18    2
2016-04-02 21:46:17    2
2016-03-08 11:55:48    2
2016-03-24 23:57:44    2
2016-03-15 21:36:44    2
2016-03-30 17:44:44    2
2016-03-24 10:56:50    2
2016-03-30 19:43:27    2
                      ..
2016-03-19 16:52:30    1
2016-03-21 20:36:19    1
2016-03-07 11:50:19    1
2016-03-27 19:54:51    1
2016-03-16 11:47:29    1
2016-03-12 12:58:02    1
2016-03-14 16:06:22    1
2016-03-27 18:56:46    1
2016-03-28 17:55:05    1


Observations:
* The columns `seller ` and `offer_type` have the same value, so they can be discarded from our analysis.
* The columns `vehicle_type`,`gearbox`,`fuel_type` and `unrepaired_damage` are in German
* All the values of the column `num_photos` are 0
* There are 1421 entries with a value of $0 in the column `price`
* The columns `price` and `odometer` are stored as strings instead of integers.


Based on our observations: 
* We will drop the following columns: `seller`,`offer_type` and `num_photos`

In [276]:
autos.drop(['offer_type','seller','num_photos'],axis=1)

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

* We will convert the columns `price` and `odometer` from string values to numeric ones

In [277]:
autos['price']=(autos['price']
                        .str.replace('$','')
                        .str.replace(',','')
                        .astype(int))
autos['odometer']=(autos['odometer']
                            .str.replace('km','')
                            .str.replace(',','')
                            .astype(int))

In [278]:
autos.rename(columns={'odometer':'odometer_km'},inplace=True)

### Exploring the Odometer and Price Columns

We will take a deep look at the ``Odometer`` and ``Price`` columns to visualize if there are any unusual values.

In [279]:
autos['price'].unique().shape

(2357,)

In [280]:
autos['odometer_km'].unique().shape

(13,)

In [281]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [282]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [283]:
    autos['price'].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

In [284]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

From this analysis we can identify that there are unrealistic prices for cars in the `price` column such as: 
* A car priced at `$99999999` or `$12345678`
* 1421 cars with a price of ``$0``

For this reason we will remove them. 
Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

For the `odometer_km` column the values seem to be within normal parameters

In [285]:
autos=autos[autos['price'].between(1,350001)]

In [286]:
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

### Exploring date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

In [287]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [288]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
6        2016-03-20
7        2016-03-16
8        2016-03-22
9        2016-03-16
10       2016-03-15
11       2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
16       2016-03-16
17       2016-03-29
18       2016-03-26
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
25       2016-03-21
26       2016-04-03
28       2016-03-19
29       2016-04-02
30       2016-03-14
            ...    
49968    2016-04-01
49969    2016-03-17
49970    2016-03-21
49971    2016-03-29
49972    2016-03-26
49973    2016-03-27
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49978    2016-04-04
49979    2016-03-20
49980    2016-03-12
49981    2016-03-15
49982    2016-03-29
49983    2016-03-06
49985    2016-04-02
49986    2016-04-04
49987    2016-03-22
49988    2016-03-28


In [289]:
print(autos['ad_created'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
6        2016-03-20
7        2016-03-16
8        2016-03-22
9        2016-03-16
10       2016-03-14
11       2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
16       2016-03-16
17       2016-03-29
18       2016-03-26
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
25       2016-03-21
26       2016-04-03
28       2016-03-19
29       2016-04-02
30       2016-03-14
            ...    
49968    2016-04-01
49969    2016-03-17
49970    2016-03-21
49971    2016-03-29
49972    2016-03-26
49973    2016-03-27
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49978    2016-04-04
49979    2016-03-20
49980    2016-03-12
49981    2016-03-15
49982    2016-03-29
49983    2016-03-06
49985    2016-04-02
49986    2016-04-04
49987    2016-03-22
49988    2016-03-28


In [290]:
print(autos['last_seen'].str[:10])

0        2016-04-06
1        2016-04-06
2        2016-04-06
3        2016-03-15
4        2016-04-01
5        2016-04-06
6        2016-03-23
7        2016-04-07
8        2016-03-26
9        2016-04-06
10       2016-04-07
11       2016-03-16
12       2016-04-06
13       2016-04-01
14       2016-04-01
15       2016-04-02
16       2016-03-18
17       2016-04-05
18       2016-04-02
19       2016-03-17
20       2016-03-07
21       2016-04-05
22       2016-04-01
23       2016-03-17
24       2016-04-07
25       2016-04-01
26       2016-04-05
28       2016-04-07
29       2016-04-02
30       2016-03-14
            ...    
49968    2016-04-05
49969    2016-03-25
49970    2016-04-07
49971    2016-03-29
49972    2016-03-28
49973    2016-03-27
49975    2016-04-05
49976    2016-04-07
49977    2016-04-06
49978    2016-04-06
49979    2016-04-07
49980    2016-04-06
49981    2016-04-05
49982    2016-03-29
49983    2016-03-09
49985    2016-04-06
49986    2016-04-06
49987    2016-03-26
49988    2016-04-07


In [291]:
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05 14:06:30    0.000021
2016-03-05 14:06:40    0.000021
2016-03-05 14:07:04    0.000021
2016-03-05 14:07:08    0.000021
2016-03-05 14:07:21    0.000021
2016-03-05 14:07:26    0.000021
2016-03-05 14:07:40    0.000021
2016-03-05 14:07:45    0.000021
2016-03-05 14:08:00    0.000041
2016-03-05 14:08:05    0.000041
2016-03-05 14:08:27    0.000021
2016-03-05 14:08:42    0.000021
2016-03-05 14:09:02    0.000041
2016-03-05 14:09:05    0.000021
2016-03-05 14:09:20    0.000021
2016-03-05 14:09:22    0.000021
2016-03-05 14:09:38    0.000021
2016-03-05 14:09:46    0.000021
2016-03-05 14:09:56    0.000021
2016-03-05 14:09:57    0.000021
2016-03-05 14:09:58    0.000041
2016-03-05 14:10:18    0.000021
2016-03-05 14:10:20    0.000021
2016-03-05 14:10:46    0.000021
2016-03-05 14:11:03    0.000021
2016-03-05 14:11:05    0.000021
2016-03-05 14:11:14    0.000021
2016-03-05 14:11:15    0.000021
2016-03-05 14:11:25    0.000021
2016-03-05 14:11:40    0.000021
                         ...   
2016-04-

Looks like that the data was crawled during the months of March and April of 2016. The distribution of listings crawled on each day is roughly uniform.

In [292]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11 00:00:00    0.000021
2015-08-10 00:00:00    0.000021
2015-09-09 00:00:00    0.000021
2015-11-10 00:00:00    0.000021
2015-12-05 00:00:00    0.000021
2015-12-30 00:00:00    0.000021
2016-01-03 00:00:00    0.000021
2016-01-07 00:00:00    0.000021
2016-01-10 00:00:00    0.000041
2016-01-13 00:00:00    0.000021
2016-01-14 00:00:00    0.000021
2016-01-16 00:00:00    0.000021
2016-01-22 00:00:00    0.000021
2016-01-27 00:00:00    0.000062
2016-01-29 00:00:00    0.000021
2016-02-01 00:00:00    0.000021
2016-02-02 00:00:00    0.000041
2016-02-05 00:00:00    0.000041
2016-02-07 00:00:00    0.000021
2016-02-08 00:00:00    0.000021
2016-02-09 00:00:00    0.000021
2016-02-11 00:00:00    0.000021
2016-02-12 00:00:00    0.000041
2016-02-14 00:00:00    0.000041
2016-02-16 00:00:00    0.000021
2016-02-17 00:00:00    0.000021
2016-02-18 00:00:00    0.000041
2016-02-19 00:00:00    0.000062
2016-02-20 00:00:00    0.000041
2016-02-21 00:00:00    0.000062
                         ...   
2016-03-

It looks like that the ads were created from June of 2015 to March of 2016.  The distribution of ads created is not uniform

In [293]:
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05 14:45:46    0.000021
2016-03-05 14:46:02    0.000021
2016-03-05 14:49:34    0.000021
2016-03-05 15:16:11    0.000021
2016-03-05 15:16:47    0.000021
2016-03-05 15:28:10    0.000021
2016-03-05 15:41:30    0.000021
2016-03-05 15:45:43    0.000021
2016-03-05 15:47:38    0.000021
2016-03-05 15:47:44    0.000021
2016-03-05 16:45:57    0.000021
2016-03-05 16:47:28    0.000021
2016-03-05 17:15:45    0.000021
2016-03-05 17:16:14    0.000021
2016-03-05 17:16:23    0.000021
2016-03-05 17:17:02    0.000021
2016-03-05 17:39:19    0.000021
2016-03-05 17:40:14    0.000021
2016-03-05 17:44:50    0.000021
2016-03-05 17:44:54    0.000021
2016-03-05 17:46:01    0.000021
2016-03-05 18:17:58    0.000021
2016-03-05 18:50:38    0.000021
2016-03-05 19:15:08    0.000021
2016-03-05 19:15:20    0.000021
2016-03-05 19:15:42    0.000021
2016-03-05 19:16:36    0.000021
2016-03-05 19:17:17    0.000021
2016-03-05 19:17:50    0.000021
2016-03-05 19:32:34    0.000021
                         ...   
2016-04-

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.



In [294]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [295]:
autos['registration_year'].value_counts()

2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
2006    2670
2001    2636
2002    2486
1998    2363
2007    2277
2008    2215
2009    2085
1997    1951
2011    1623
2010    1589
2017    1392
1996    1373
2012    1310
1995    1227
2016    1220
2013     803
2014     663
1994     629
2018     470
1993     425
2015     392
1992     370
1990     347
1991     339
1989     174
        ... 
1950       3
9999       3
1800       2
1951       2
1941       2
1934       2
1954       2
1957       2
1955       2
2019       2
1001       1
1939       1
4500       1
1953       1
1111       1
4800       1
5911       1
1943       1
1938       1
1929       1
2800       1
6200       1
4100       1
8888       1
1927       1
9000       1
1948       1
1000       1
1931       1
1952       1
Name: registration_year, Length: 95, dtype: int64

We can clearly notice that some registration years are definitely wrong 
E.g: 4500 - 1111 - 1000 - 8888 - 4800 - 2800 - 9000 
The minimun value of a year is 1000 and the maximun one is 9999 
We will remove any value that falls outside the 1900 - 2016 interval.

### Exploring Price by Brand

In [296]:
autos=autos[autos['registration_year'].between(1900,2016)]

In [297]:
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

We narrowed the values from 48565 to 46681. We removed 1884 entries.

In [298]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

In [301]:
autos_brand=autos['brand'].value_counts(normalize=True)
top_brands=autos_brand[autos_brand > .05].index #index is important for the next step
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


We are exploring the top car companies that are listed on eBay Germany. The 50% of the market share is made up from German companies, with VW leading the chart 
with 21%.

In [313]:
top_autos_price={}
for c in top_brands:
    selected_brands=autos[autos['brand']==c]
    means=selected_brands['price'].mean()
    top_autos_price[c]=int(means)
top_autos_price

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}


Of the top 5 brands, there is a distinct price gap:

Audi, BMW and Mercedes Benz are more expensive
Ford and Opel are less expensive
Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

### Exploring Aggregate Data in DataFrame

In [311]:
top_mean_mileage={}
for c in top_brands:
    selected_brands=autos[autos['brand']==c]
    means=selected_brands['odometer_km'].mean()
    top_mean_mileage[c]=means
top_mean_mileage

{'audi': 129157.38678544914,
 'bmw': 132572.51313996495,
 'ford': 124266.01287159056,
 'mercedes_benz': 130788.36331334666,
 'opel': 129310.0358422939,
 'volkswagen': 128707.15879132022}

In [318]:
tap_series = pd.Series(top_autos_price)
tap=pd.DataFrame(tap_series, columns=["mean_price"])
tap

mean_price
audi                 9336
bmw                  8332
ford                 3749
mercedes_benz        8628
opel                 2975
volkswagen           5402

In [319]:
tmm_series = pd.Series(top_mean_mileage)
tmm=pd.DataFrame(tmm_series, columns=["mean_mileage"])
tmm

mean_mileage
audi           129157.386785
bmw            132572.513140
ford           124266.012872
mercedes_benz  130788.363313
opel           129310.035842
volkswagen     128707.158791

In [320]:
tmm["mean_price"] = tap

In [321]:
tmm

mean_mileage  mean_price
audi           129157.386785        9336
bmw            132572.513140        8332
ford           124266.012872        3749
mercedes_benz  130788.363313        8628
opel           129310.035842        2975
volkswagen     128707.158791        5402


The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.